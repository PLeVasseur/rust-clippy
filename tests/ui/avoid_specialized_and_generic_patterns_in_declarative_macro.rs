#![warn(clippy::avoid_specialized_and_generic_patterns_in_declarative_macro)]

#[derive(Debug)]
enum SafetyLevel {
    Green,
    Yellow,
    Red,
}
// Different device types that need safety checks
struct PressureSensor {/* ... */}
struct TemperatureSensor {/* ... */}
struct EmergencyValve {
    open: bool,
}

trait SafetyCheckForSpecializationIssue {
    fn verify(&self) -> SafetyLevel;
}
// This macro has a pattern ordering issue possible on refactor
macro_rules! impl_safety_trait_specialization_issue {
    // Generic pattern matches any type - including EmergencyValve
    ($u:ty) => {
        impl SafetyCheckForSpecializationIssue for $u {
            fn verify(&self) -> SafetyLevel {
                SafetyLevel::Green
            }
        }
    };
    // Generic pattern matches any type - including EmergencyValve
    ($t:ty, EmergencyValve) => {
        impl SafetyCheckForSpecializationIssue for $t {
            fn verify(&self) -> SafetyLevel {
                SafetyLevel::Green
            }
        }
    };
    // Special pattern for EmergencyValve - but never gets matched
    (EmergencyValve) => {
        //~^ avoid_specialized_and_generic_patterns_in_declarative_macro
        impl SafetyCheckForSpecializationIssue for EmergencyValve {
            fn verify(&self) -> SafetyLevel {
                // Emergency valve must be open for safety
                if !self.open {
                    SafetyLevel::Red
                } else {
                    SafetyLevel::Green
                }
            }
        }
    };
}
impl_safety_trait_specialization_issue!(EmergencyValve);
impl_safety_trait_specialization_issue!(PressureSensor);
impl_safety_trait_specialization_issue!(TemperatureSensor);

trait SafetyCheckWithoutSpecializationIssue {
    fn verify(&self) -> SafetyLevel;
}
// This macro has a pattern ordering issue possible on refactor
macro_rules! impl_safety_trait_without_specialization_issue {
    // Generic pattern matches any type - including EmergencyValve
    ($t:ty) => {
        impl SafetyCheckWithoutSpecializationIssue for $t {
            fn verify(&self) -> SafetyLevel {
                SafetyLevel::Green
            }
        }
    };
}
impl_safety_trait_without_specialization_issue!(PressureSensor);
impl_safety_trait_without_specialization_issue!(TemperatureSensor);
impl SafetyCheckWithoutSpecializationIssue for EmergencyValve {
    fn verify(&self) -> SafetyLevel {
        // Emergency valve must be open for safety
        if !self.open {
            SafetyLevel::Red
        } else {
            SafetyLevel::Green
        }
    }
}

fn main() {
    // test code goes here
}
