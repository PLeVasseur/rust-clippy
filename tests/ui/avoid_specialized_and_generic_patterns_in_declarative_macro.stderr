error: avoid specialized matchers in declarative macros alongside generic ones
  --> tests/ui/avoid_specialized_and_generic_patterns_in_declarative_macro.rs:20:53
   |
LL | macro_rules! impl_safety_trait_specialization_issue {
   |                                                     ^
   |
   = help: consider implementing specialized patterns outside of macro
   = note: `-D clippy::avoid-specialized-and-generic-patterns-in-declarative-macro` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::avoid_specialized_and_generic_patterns_in_declarative_macro)]`

error: avoid specialized matchers in declarative macros alongside generic ones
  --> tests/ui/avoid_specialized_and_generic_patterns_in_declarative_macro.rs:52:61
   |
LL | macro_rules! impl_safety_trait_without_specialization_issue {
   |                                                             ^
   |
   = help: consider implementing specialized patterns outside of macro

error: aborting due to 2 previous errors

